Mathlib/GroupTheory/Coxeter/Basic.lean

```diff
@@ -0,0 +1,695 @@
+  exact hf i i'
+
+private def groupLift {G : Type*} [Group G] {f : B → G} (hf : IsLiftable M f) : W →* G :=
+  (PresentedGroup.toGroup (relations_liftable hf)).comp cs.mulEquiv.toMonoidHom
+
+private def restrictUnit {G : Type*} [Monoid G] {f : B → G} (hf : IsLiftable M f) (i : B) :
+    Gˣ where
+  val := f i
+  inv := f i
+  val_inv := pow_one (f i * f i) ▸ (cs.isCoxeter.diagonal i ▸ hf i i)
+  inv_val := pow_one (f i * f i) ▸ (cs.isCoxeter.diagonal i ▸ hf i i)
+
+/-- Extend the function `f : B → G` to a monoid homomorphism
+`f' : W → G` satisfying `f' (s i) = f i` for all `i`.
+-/
+def lift {G : Type*} [Monoid G] {f : B → G} (hf : IsLiftable M f) : W →* G :=
+  MonoidHom.comp (Units.coeHom G) (cs.groupLift
+    (show ∀ i i' : B, ((cs.restrictUnit hf) i * (cs.restrictUnit hf) i') ^ M i i' = 1 from
+      fun i i' ↦ Units.ext (hf i i')))
+
+private theorem toMonoidHom_symm (a : PresentedGroup (CoxeterGroup.Relations.toSet M)):
+    (MulEquiv.toMonoidHom cs.mulEquiv : W →* PresentedGroup (CoxeterGroup.Relations.toSet M))
+    ((MulEquiv.symm cs.mulEquiv) a) = a := calc
+  _ = cs.mulEquiv ((MulEquiv.symm cs.mulEquiv) a) := by rfl
+  _ = _                                           := by rw [MulEquiv.apply_symm_apply]
+
+theorem lift_apply_simple {G : Type*} [Monoid G] {f : B → G}
+    (hf : IsLiftable M f) (i : B) : cs.lift hf (s i) = f i := by
+  dsimp only [simpleReflection, lift, groupLift, MonoidHom.comp_apply]
+  rw [← MonoidHom.toFun_eq_coe, toMonoidHom_symm cs (PresentedGroup.of i)]
+  simp
+  rfl
+
+/-- If two homomorphisms with domain `W` agree on all simple reflections, then they are equal. -/
+theorem ext_simple {G : Type*} [Monoid G] {φ₁ φ₂ : W →* G} (h : ∀ i : B, φ₁ (s i) = φ₂ (s i)) :
+    φ₁ = φ₂ := by
+  apply MonoidHom.eq_of_eqOn_denseM (cs.submonoid_closure_range_simple)
+  rintro x ⟨i, rfl⟩
+  exact h i
+
+/-- If two Coxeter systems on the same group `W` have the same Coxeter matrix `M : Matrix B B ℕ`
+and the same simple reflection map `B → W`, then they are identical. -/
+theorem simpleReflection_determines_coxeterSystem :
+    Injective (simpleReflection : CoxeterSystem M W → B → W) := by
+  intro cs1 cs2 h
+  apply CoxeterSystem.ext
+  apply MulEquiv.toMonoidHom_injective
+  apply cs1.ext_simple
+  intro i
+  nth_rw 2 [h]
+  simp [simpleReflection]
+
+/-! ### Words -/
+
+/-- The product of the simple reflections of `W` corresponding to the indices in `ω`.-/
+def wordProd (ω : List B) : W := prod (map cs.simpleReflection ω)
+
+local prefix:100 "π" => cs.wordProd
+
+@[simp] theorem wordProd_nil : π [] = 1 := by simp [wordProd]
+
+@[simp] theorem wordProd_cons (i : B) (ω : List B) : π (i :: ω) = s i * π ω := by simp [wordProd]
+
+theorem wordProd_singleton (i : B) : π ([i]) = s i := by simp [wordProd]
+
+theorem wordProd_concat (i : B) (ω : List B) : π (ω.concat i) = π ω * s i := by simp [wordProd]
+
+@[simp] theorem wordProd_append (ω ω' : List B) : π (ω ++ ω') = π ω * π ω' := by simp [wordProd]
```

Comment:
This lemma is not "affine": There are variables (namely `cs`) with more occurrences on the RHS than on the RHS. This means that if `cs` is simplifiable then you could hit exponential blowup:
`π (a ++ b ++ c ++ d ++ e) → π a * π b * π c * π d * π e → simplified π a * simplified π b * simplified π c * simplified π d * simplified π e`.

It is therefore advisable for this not to be a simp lemma.